backend/src/main/java/org/eclipse/tractusx/puris/backend/common/edc/logic/service/EdcAdapterService.java

```diff
@@ -487,6 +541,420 @@ public Response getProxyPullRequest(String url, String authKey, String authCode,
         }
     }
 
+    private boolean getPartnerAasForMaterial(Partner partner, String partnerMaterialNumber) {
+        EdcContractMapping contractMapping = edcContractMappingService.find(partner.getBpnl());
+        if (contractMapping != null) {
+            if (contractMapping.getDtrContractId() == null || contractMapping.getDtrAssetId() == null) {
+                negotiateForPartnerDtr(partner);
+            }
+            return initiateDtrTransferForLookupApi(partner, "manufacturerPartId", partnerMaterialNumber);
+        }
+        return false;
+    }
+
+    private boolean getPartnerAasForProduct(Partner partner, String partnerMaterialNumber) {
+        EdcContractMapping contractMapping = edcContractMappingService.find(partner.getBpnl());
+        if (contractMapping != null) {
+            if (contractMapping.getDtrContractId() == null || contractMapping.getDtrAssetId() == null) {
+                negotiateForPartnerDtr(partner);
+            }
+            return initiateDtrTransferForLookupApi(partner, "customerPartId", partnerMaterialNumber);
+        }
+        return false;
+    }
+
+    public JsonNode doItemStockSubmodelRequest(MaterialPartnerRelation mpr, DirectionCharacteristic direction) {
+        Partner partner = mpr.getPartner();
+        Material material = mpr.getMaterial();
+        String materialNumber = switch (direction) {
+            case INBOUND -> material.getMaterialNumberCx();
+            case OUTBOUND -> mpr.getPartnerCXNumber();
+        };
+        boolean failed = true;
+        EdcContractMapping contractMapping = edcContractMappingService.find(partner.getBpnl());
+        try {
+            if (contractMapping != null) {
+                String href = contractMapping.getMaterialToHrefMapping().get(materialNumber);
```

Comment:
I think the href only on material number level does not work. If I'm a supplier who supplies multiple customers, each customer - despite having the same cxId, as they copy mine - can still have different hrefs as those can be specific to the partner's implementation.
For that case a key materialNumber + bpnl would be bullet-proof.

Reply:
Done. I solved this via hand-made Getter- and Setter-methods on the EdcContractMapping entity.